providers/implementations/keymgmt/rsa_kmgmt.c

```diff
@@ -162,30 +168,85 @@ static int key_to_params(RSA *rsa, OSSL_PARAM_BLD *tmpl)
     return ret;
 }
 
-static void *rsa_importkey(void *provctx, const OSSL_PARAM params[])
+static void *rsa_newdata(void *provctx)
 {
-    RSA *rsa;
+    return RSA_new();
+}
 
-    if ((rsa = RSA_new()) == NULL
-        || !params_to_key(rsa, params)) {
-        RSA_free(rsa);
-        rsa = NULL;
-    }
-    return rsa;
+static void rsa_freedata(void *keydata)
+{
+    RSA_free(keydata);
 }
 
-static int rsa_exportkey(void *key, OSSL_CALLBACK *param_callback, void *cbarg)
+static int rsa_has_public(void *keydata)
 {
-    RSA *rsa = key;
+    RSA *rsa = keydata;
+
+    return (RSA_get0_n(rsa) != NULL && RSA_get0_e(rsa) != NULL);
+}
+
+static int rsa_has_private(void *keydata)
+{
+    RSA *rsa = keydata;
+
+    return (RSA_get0_d(rsa) != NULL && RSA_get0_e(rsa) != NULL);
+}
+
+static int rsa_import(void *keydata, int selection, const OSSL_PARAM params[])
+{
+    RSA *rsa = keydata;
+
+    if (rsa == NULL)
+        return 0;
+
+    /* TODO(3.0) PSS and OAEP should bring on parameters */
+    /* If the caller wants domain parameters, we do nothing for the moment */
+    if (selection == OSSL_KEYMGMT_WANT_DOMPARAMS)
```

Comment:
Just a note that : PSS and OAEP are NOT and SHOULD NOT EVER be treated as domain parameters. It would be very confusing.

Reply:
What should they be treated as, then?  Maybe it's just about wordings, should I call them "attributes" (mimicking the PKCS#8 terminology)?  I would have no remorse whatsoever making that kind of name change if that makes more sense.

Do note that `crypto/rsa/rsa_ameth.c` talks about them in terms of parameters...  is it "domain" specifically that's disturbing?

Reply:
Or, seeing your other comment, "properties" work too.  Can domain parameters be called properties too, or is that like swearing in church?

Reply:
Yep. RSA has no domain parameters.
Maybe making it bit flags and treating them differently would be better.
i.e have domain params, and some other params..

Reply:
Arent those things you are talking about just padding related?

Reply:
> Arent those things you are talking about just padding related?

Not in the case of PSS

Reply:
Changes have been done to allow the distinction while still maintaining the flexibility that the current EVP_PKEY functionality requires